[Test gap] Add coverage for IPv6 traffic to the PFCWD PTF test cases

peterbailey-arista opened the pull request:
Add IPv6 test coverage for PFCWD.

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/sonic-net/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->

Summary:
Fixes # (issue)
https://github.com/sonic-net/sonic-mgmt/issues/10959

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] New Test case
    - [ ] Skipped for non-supported platforms
- [ ] Test case improvement


### Back port request
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [ ] 202405
- [ ] 202411

### Approach
#### What is the motivation for this PR?
https://github.com/sonic-net/sonic-buildimage/issues/17446 
This issue was raised regarding PFCWD not dropping IPv6 traffic during storm. It revealed the need for test coverage.

#### How did you do it?
- I added a parametrized ip_version fixture [*tests/conftest.py*]
- This fixture is used in the setup_pfc_test fixture [*tests/conftest.py*]
- All the tests that use setup_pfc_test to initialize the TrafficPorts will create them using once with IPv4 and once with IPv6
- TrafficPorts [*tests/common/helpers/pfcwd_helper.py*] uses the ip version to determine the test interfaces' and neighbour's ip addresses
- The ip_version is used explicitly in tests with a SendVerifyTraffic class [*tests/pfcwd/test_pfcwd_cli.py*, *tests/pfcwd/test_pfcwd_warm_reboot.py*, *tests/pfcwd/test_pfcwd_function.py*] that calls the ptf_runner with PfcWdTest [*ansible/roles/test/files/ptftests/py3/pfc_wd.py*]
  - We use the IP version to properly generate addresses and to properly create tcp packets in PfcWdTest
- I updated pfcwd timer accuracy test cleanup [*tests/pfcwd/test_pfcwd_timer_accuracy.py*] to perform a config reload, this was required to clear the `show pfcwd stats` in between the IPv4 and IPv6 runs
  - it also removed the need for `stop_pfcwd` fixture as the pfcwd is already stopped by setup_pfc_test, the `stop_pfcwd` fixture additionally restarted the wd (https://github.com/sonic-net/sonic-mgmt/pull/15025) but this is now unnecessary due to the config reload.
- Changes for VLAN compatibility:
  - resolve_arp [*tests/pfcwd/test_pfcwd_cli.py*, *tests/pfcwd/test_pfcwd_warm_reboot.py*, *tests/pfcwd/test_pfcwd_function.py*] required IPv6 compatible bash commands
  - get_ip_in_range [*ansible/library/get_ip_in_range.py*] required refactoring to allow for lazy loading of large address spaces (i.e. IPv6 spaces)

#### How did you verify/test it?
Ran all pfcwd sonic mgmt tests across T0, T1, and T2 topologies all with successful results.

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/AzurePipelines run Azure.sonic-mgmt

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

peterbailey-arista commented:
kvmtest-t0 failure caused by issue https://github.com/sonic-net/sonic-mgmt/issues/17803

vmittal-msft commented:
Please rebase to update. Also, have we verified on T0 as well as T2 device ?

abdosi commented:
@kenneth-arista will check we don;t see same regression as last time that needed PR to be reverted.

peterbailey-arista commented:
Update: I am actively working on resolving flakiness in T0 test cases.

kenneth-arista commented:
> @kenneth-arista will check we don;t see same regression as last time that needed PR to be reverted.

@abdosi The previous PR, which was reverted, was in Qos tests and not in PFCWD which this PR is about. Peter will do the due diligence to confirm that these changes will work across different topos.

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

peterbailey-arista commented:
/azpw run Azure.sonic-mgmt

mssonicbld commented:
/AzurePipelines run Azure.sonic-mgmt

peterbailey-arista commented:
The updated cases pass all tests except the issue described by https://github.com/sonic-net/sonic-mgmt/issues/17803 has continued to persist. I believe this is unrelated to my changes.

vmittal-msft commented:
/azp run

mssonicbld commented:
/AzurePipelines run Azure.sonic-mgmt

peterbailey-arista commented:
@vmittal-msft I don't believe the last failing test is related to my additions, it is only the `pfcwd/test_pfcwd_warm_reboot.py` `async-storm` test case failing and this test has been failing with the same `Exception: warmboot-finalizer service timeout on DUT vlab-01` error since before this change. (As seen in https://github.com/sonic-net/sonic-mgmt/issues/17803)

vmittal-msft commented:
@peterbailey-arista please rebase and restart pipeline. if still fails, will follow up.

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

peterbailey-arista commented:
> @peterbailey-arista please rebase and restart pipeline. if still fails, will follow up.

Hi @vmittal-msft , I rebased and updated my tests to work with the new changes. Still the same single test fails
(https://github.com/sonic-net/sonic-mgmt/pull/18247#issuecomment-2985744094)

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/AzurePipelines run Azure.sonic-mgmt

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

peterbailey-arista commented:
@vmittal-msft I have all the pipeline stages passing now 😃

vmittal-msft commented:
Thanks @peterbailey-arista for your hard work for making this work. I could see that you had to go through lot of iterations for this.  @rlhui @yxieca please help merge this.
